R/merge_sce_list.R

```diff
@@ -424,3 +429,138 @@ build_new_altexp_assay <- function(
+
+    # We only want to work with rowDatas which contain this colname of interest,
+    # for dplyr::select appeasement
+    rowdata_list_with_colname <- rowdata_list |>
+      purrr::map(
+        \(rowdata) {
+          if (colname %in% colnames(rowdata)) {
+            return(rowdata)
+          } else {
+            return(NULL)
+          }
+        }
+      ) |>
+      purrr::discard(is.null)
+
+    if (length(rowdata_list_with_colname) == 0) {
+      next
+    }
+    colname_df <- rowdata_list_with_colname |>
+      # Subset to only this `colname`, with rownames stored in `temp` for joining
+      purrr::map(
+        \(rowdata) {
+            rowdata |>
+              dplyr::select(all_of(colname)) |>
+              # pull out temp for joining
+              tibble::rownames_to_column("temp")
+        }
+      ) |>
+      # combine all versions of `colname` into one data frame
+      # this way, we ensure that we consider _all features_
+      purrr::reduce(dplyr::full_join, by = "temp") |>
+      # back to data frame with rownames, and drop temp
+      as.data.frame()
+    rownames(colname_df) <- colname_df$temp
+    colname_df$temp <- NULL
+
+    # The column with the fewest NAs is what we want to use
+    keep_column_name <- colname_df |>
+      is.na() |>
+      colSums() |>
+      # which is the minimum? If ties, this returns the first occurrence only.
+      which.min() |>
+      names()
+
+    # Save for later use
+    preserved_columns[[colname]] <- colname_df |>
+      # {{}} and := set the colname to the actual value
+      dplyr::select({{colname}} := all_of(keep_column_name))
+  }
+  # Convert to data frame for easily adding into the rowData later
+  preserved_columns_df <- purrr::list_cbind(preserved_columns)
+
+
+
+  new_rowdata <- rowdata_list |>
+    # Remove all of the `preserve_rowdata_cols`
+    purrr::map(
+      \(rowdata) {
+        if (colname %in% colnames(rowdata)) {
+          return(rowdata |> dplyr::select(-all_of(preserve_rowdata_cols)))
+        } else {
+          return(rowdata)
+        }
+      }
+    ) |>
+    # Rename remaining columns as `{sce_name}-{column}`
+    purrr::imap(
+      \(rowdata, sce_name) {
+        rowdata |>
+          purrr::set_names( glue::glue("{sce_name}-{names(rowdata)}") )
+      }
+    ) |>
+    # Next, ensure all rows are present. For any missing rows, add all `NA` values
+    purrr::map(
+      \(rowdata) {
+        # find the features in this rowdata
+        present_features <- rownames(rowdata)
+
+        # how many are missing?
+        missing_features <- setdiff(all_merged_features, present_features)
+
+        # add this number of new rows with all NA values
+        if (length(missing_features) > 0) {
+          rowdata[nrow(rowdata) + length(missing_features),] <- NA
+        }
+
+        # name those rows
+        rownames(rowdata) <- c(present_features, missing_features)
+
+        # return updated rowdata
+        rowdata
```

Comment:
Note that this will not be in the same order as `all_merged_features`, which is a problem.